packages/ai/src/ui/abstract-chat.ts

```diff
@@ -101,6 +69,59 @@ export interface ChatState<MESSAGE_METADATA, DATA_TYPES extends UIDataTypes> {
   snapshot?: <T>(thing: T) => T;
 }
 
+export interface AbstractChatInit<
+  MESSAGE_METADATA = unknown,
+  UI_DATA_PART_SCHEMAS extends UIDataPartSchemas = UIDataPartSchemas,
+> {
+  id: string;
+
+  /**
+   * A way to provide a function that is going to be used for ids for messages and the chat.
+   * If not provided the default AI SDK `generateId` is used.
+   */
+  generateId?: IdGenerator;
+  transport: ChatTransport<
+    MESSAGE_METADATA,
+    InferUIDataParts<UI_DATA_PART_SCHEMAS>
+  >;
+  maxSteps?: number;
+  messageMetadataSchema?:
+    | Validator<MESSAGE_METADATA>
+    | StandardSchemaV1<MESSAGE_METADATA>;
+  dataPartSchemas?: UI_DATA_PART_SCHEMAS;
+  state: ChatState<MESSAGE_METADATA, InferUIDataParts<UI_DATA_PART_SCHEMAS>>;
+
+  /**
+   * Callback function to be called when an error is encountered.
+   */
+  onError?: (error: Error) => void;
+
+  /**
+  Optional callback function that is invoked when a tool call is received.
+  Intended for automatic client-side tool execution.
+  
+  You can optionally return a result for the tool call,
+  either synchronously or asynchronously.
+     */
+  onToolCall?: ({
+    toolCall,
+  }: {
+    toolCall: ToolCall<string, unknown>;
+  }) => void | Promise<unknown> | unknown;
+
+  /**
+   * Optional callback function that is called when the assistant message is finished streaming.
+   *
+   * @param message The message that was streamed.
+   */
+  onFinish?: (options: {
+    message: UIMessage<
+      MESSAGE_METADATA,
+      InferUIDataParts<UI_DATA_PART_SCHEMAS>
+    >;
+  }) => void;
+}
```

Comment:
Pushed a bunch of stuff from the library implementations back to here because it makes more sense and deduplicates a bunch of code